[SPARK-5311][core] Corrected EventLoggingListener throws exception if log directory does not exist

ganonp opened the pull request:
Corrected exception handling: 

Now instead of blocking execution, when the baselogpath refers to a nonexistant directory, one is created at the baselogpath location. As intended, if the baselogpath is a file and not a directory, an exception is thrown.

ganonp committed:
Added setMinCount method to Word2Vec

Converted minCount from val to var and created setMinCount method.

Allows one to set the minimum number of times a word must be present to
be included into the vocabulary.

ganonp committed:
Revert "Added setMinCount method to Word2Vec"

This reverts commit 1867869f5dbc19584e9f2827c440bd0134ab84e9.

ganonp committed:
update

ganonp committed:
Merge branch 'master' of https://github.com/apache/spark into apache-master

ganonp committed:
Corrects Issue SPARK-5311

Issue Description: EventLoggingListener throws exception if log directory does not exist

Replaced with code that uses a try catch with getFileStatus method to first determine if the file exists (if it doesn't it is created in the catch block), if it does exist, then an exception is thrown if the file is not a directory.

ganonp committed:
Update EventLoggingListener.scala

ganonp committed:
Update EventLoggingListener.scala

ganonp committed:
Update EventLoggingListener.scala

ganonp committed:
Update EventLoggingListener.scala

AmplabJenkins commented:
Can one of the admins verify this patch?

CodingCat commented on core/src/main/scala/org/apache/spark/scheduler/EventLoggingListener.scala:
```diff
@@ -86,8 +86,14 @@ private[spark] class EventLoggingListener(
+        if(!fileSystem.getFileStatus(new Path(logBaseDir)).isDir()){
+          throw new IllegalArgumentException(s"Log directory $logBaseDir is not a directory.")
+        }
+        
```
this is an extra line?

CodingCat commented on core/src/main/scala/org/apache/spark/scheduler/EventLoggingListener.scala:
```diff
@@ -86,8 +86,14 @@ private[spark] class EventLoggingListener(
   def start() {
-    if (!fileSystem.isDirectory(new Path(logBaseDir))) {
-      throw new IllegalArgumentException(s"Log directory $logBaseDir does not exist.")
+    try(
```
I think they prefer {} to ()

CodingCat commented:
it seems that the change of the mode of make-distribution.sh is introduced by accident...we need recover that?

ganonp committed:
creates new directory on correct filesystem

ganonp committed:
Changed () to {} after try

vanzin commented on core/src/main/scala/org/apache/spark/scheduler/EventLoggingListener.scala:
```diff
@@ -86,8 +86,14 @@ private[spark] class EventLoggingListener(
   def start() {
-    if (!fileSystem.isDirectory(new Path(logBaseDir))) {
-      throw new IllegalArgumentException(s"Log directory $logBaseDir does not exist.")
+    try{
```
nit: you needs spaces after `{` and `(` in all your code.

ganonp replied:
Do I need them before the closing } and ) as well?

JoshRosen replied:
Here's our code style guide: https://cwiki.apache.org/confluence/display/SPARK/Spark+Code+Style+Guide

The Spark style would format the code like this:

```scala
try {
  if (!filesystem.getFileStatus(newPath(logBaseDir)).isDir()) {
    throw new IllegalArgumentException(s"Log directory $logBaseDir is not a directory.")
  }
} catch {
  case ioe: FileNotFoundException => fileSystem.mkdirs(newPath(logBaseDir)))
}
```

ganonp replied:
Ok, got it, sorry and thanks for correcting it. I read the style guid last time I contributed, but obviously forgot and didn't look at it again this time. However, I did use the scalastyle script in the dev folder this time and it said everything was good, so I probably need to figure out what I did wrong with that.

vanzin commented on core/src/main/scala/org/apache/spark/scheduler/EventLoggingListener.scala:
```diff
@@ -86,8 +86,14 @@ private[spark] class EventLoggingListener(
+        }
+        
+    }
+    catch{
```
nit: this goes in the previous line.

vanzin commented:
Doesn't seem like you should be touching the `make-distribution.sh` script.

JoshRosen commented:
We should also add a regression test to EventLoggingListenerSuite; this can try to use a non-existent subdirectory of the test directory as the event log dir.

vanzin commented:
It's a little sketchy to have the app create this directory, since it will probably not do so with the desired permissions. But if that's the old behavior, well, guess it will have to do.

pwendell commented on core/src/main/scala/org/apache/spark/scheduler/EventLoggingListener.scala:
```diff
@@ -86,8 +86,14 @@ private[spark] class EventLoggingListener(
+        
+    }
+    catch{
+        case ioe: FileNotFoundException => fileSystem.mkdirs(new Path(logBaseDir))
```
Let's at log a warning before creating this since as @vanzin points out, it may not have correct permissions for downstream consumers.

ganonp committed:
Update EventLoggingListener.scala

vanzin commented on core/src/main/scala/org/apache/spark/scheduler/EventLoggingListener.scala:
```diff
@@ -86,8 +86,13 @@ private[spark] class EventLoggingListener(
-    if (!fileSystem.isDirectory(new Path(logBaseDir))) {
-      throw new IllegalArgumentException(s"Log directory $logBaseDir does not exist.")
+    try{
+        if( !fileSystem.getFileStatus( new Path( logBaseDir ) ).isDir() ){
```
This is still not right. Just take a look at the code Josh posted, or the code you're replacing. Basically, do this:

    if (foo) {

Not this:

    if( foo ){

ganonp replied:
Sorry, I'm losing my mind, I changed it to that, then I saw his, and I thought he had changed it. Should be right now.

vanzin commented:
Also, you still need to undo the mode change on `make-distribution.sh`.

ganonp committed:
Update EventLoggingListener.scala

JoshRosen commented on core/src/main/scala/org/apache/spark/scheduler/EventLoggingListener.scala:
```diff
@@ -86,8 +86,12 @@ private[spark] class EventLoggingListener(
+        throw new IllegalArgumentException(s"Log directory $logBaseDir is not a directory.")
+      }
+    } catch {
+      case ioe: FileNotFoundException => fileSystem.mkdirs(new Path(logBaseDir))
```
Per @pwendell's earlier comment, I think we should check whether `logBaseDir` already exists and log a warning if it's being automatically created to let users know that it could potentially have wider directory permissions than they intended.

ganonp commented:
Yeah, I will do that shortly. I got a little ahead of myself on this pull request, as I thought the fileSystem.getfilestatus check and fileSystem.mkdirs had resolved the problem before I had adequately tested. Unfortunately, now I'm getting new exceptions a little further down in the EventLoggingListener class. I've made some changes and am currently rebuilding spark and testing. 

In the meantime, is the proper protocol to close the pull request until I have it working correctly, or do I leave it open since it should be in working order shortly?

JoshRosen commented:
It's fine to leave the PR open if you're still planning to work on it.  Just make sure to leave a comment to ping us when you bring it up to date.

ganonp committed:
Added logWarning and File.mkDir

Hopefully this is the correct way to add logging also this takes care of a newly introduced exception. Still receiving an exception:

Exception in thread "main" java.lang.IllegalArgumentException: Wrong FS: file:/tmp/spark-events/app-20150121221900-0001.inprogress, expected: hdfs://ec2-54-191-185-61.us-west-2.compute.amazonaws.com:9000
	at org.apache.hadoop.fs.FileSystem.checkPath(FileSystem.java:381)
	at org.apache.hadoop.hdfs.DistributedFileSystem.getPathName(DistributedFileSystem.java:129)
	at org.apache.hadoop.hdfs.DistributedFileSystem.setPermission(DistributedFileSystem.java:531)
	at org.apache.spark.scheduler.EventLoggingListener.start(EventLoggingListener.scala:130)
	at org.apache.spark.SparkContext.<init>(SparkContext.scala:363)
	at okmaven.okmaven.App$.main(App.scala:80)
	at okmaven.okmaven.App.main(App.scala)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:483)
	at org.apache.spark.deploy.SparkSubmit$.launch(SparkSubmit.scala:366)
	at org.apache.spark.deploy.SparkSubmit$.main(SparkSubmit.scala:75)
	at org.apache.spark.deploy.SparkSubmit.main(SparkSubmit.scala)

ganonp commented:
Cool, thanks

JoshRosen commented:
Jenkins, this is ok to test.

SparkQA commented:
[Test build #26067 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26067/consoleFull) for   PR 4120 at commit [`c2c2001`](https://github.com/apache/spark/commit/c2c2001a856a0583bea58d5e8396da2ab95fd38e).
 * This patch merges cleanly.

SparkQA commented:
[Test build #26067 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26067/consoleFull) for   PR 4120 at commit [`c2c2001`](https://github.com/apache/spark/commit/c2c2001a856a0583bea58d5e8396da2ab95fd38e).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/26067/
Test FAILed.

JoshRosen commented:
Jenkins, retest this please.

SparkQA commented:
[Test build #26259 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26259/consoleFull) for   PR 4120 at commit [`c2c2001`](https://github.com/apache/spark/commit/c2c2001a856a0583bea58d5e8396da2ab95fd38e).
 * This patch **does not merge cleanly**.

SparkQA commented:
[Test build #26259 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26259/consoleFull) for   PR 4120 at commit [`c2c2001`](https://github.com/apache/spark/commit/c2c2001a856a0583bea58d5e8396da2ab95fd38e).
 * This patch **passes all tests**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/26259/
Test PASSed.

JoshRosen commented on core/src/main/scala/org/apache/spark/scheduler/EventLoggingListener.scala:
```diff
@@ -86,8 +86,17 @@ private[spark] class EventLoggingListener(
+      case ioe: FileNotFoundException => {
+        logWarning(s"Log directory $logBaseDir does not exist. Creating with default file " +
+          "permissions...")
+        new File(logBaseDir).mkdir()
```
We should remove this line, since we need to use `fileSystem` to create the file.  Also, I think we should use `mkdir` and not `mkdirs`, since I think that the old code failed if they didn't exist.

JoshRosen replied:
Actually, scratch that: the old behavior was to use `mkdirs`.  Sigh...

vanzin replied:
Wait. This line does need to be removed. The correct `FileSystem` call is right below.

JoshRosen replied:
Yeah, I mixed two comments into one: we need to remove this line, but I was referring to `mkdir` vs `mkdirs` in the line below.

JoshRosen commented:
Actually, maybe we don't want to fix this: the old behavior was pretty bad since it might create an entire hierarchy of directories with the default permissions.  @andrewor14, do you have any thoughts on this?

andrewor14 commented:
@ganonp I agree with @JoshRosen that we should probably not arbitrarily create the directory for the user if it doesn't exist. The semantics of the `spark.eventLog.dir` should be that "this is an existing directory and I want to use it to put my event logs". Would you mind closing this PR then?
